content/blog/2019-11-21-transmeta.md

```diff
@@ -0,0 +1,126 @@
+the paper the large instructions are "molecules" composed of 2 or 4 "atoms".
+The internal ISA avoids handling pipeline stalls--instead it expects the CMS
+compiler to generate safe code by separating conflicting operations.
+
+The hardware supports deoptimization by shadowing state and exposing `commit`
+and `rollback` operations for copying live state to the shadowed state and
+reverting to shadowed state respectively. In particular, every register has
+a corresponding shadow register. All writes to memory are held in a gated store
+buffer that is only flushed to main memory following a `commit`.
+
+## The Code Morphing System
+The CMS includes a software x86 interpreter which runs programs accurately while
+also monitoring performance statistics. Once it notices a particular code region
+has run more than some threshold number of times, it stops interpreting, commits
+the current state, and tries running a just-in-time compiled version of the code
+region. The compiled code is stored in a "translation cache" or Tcache.
+
+The JIT will reorder instructions in order to get an efficient schedule on the
+VLIW architecture. This is necessary for performance: Figure 2 in the paper
+shows a mean of 33% performance degradation across several applications when
+reordering is disabled.
+
+### Exceptions and Interrupts
+Occasionally compiled code will encounter exceptions in the internal ISA.
+Sometimes these exceptions are the result of speculative compilation (e.g.,
+a reordering of instructions causing a memory fault) but sometimes they are
+genuine exceptions which should be propagated up to the x86 layer (e.g.,
+division by zero).
+
+When Tcached code hits an exception, the CMS issues a rollback instruction to
+restore architectural state to a previous checkpoint and tries interpreting the
+region instead. If the exception goes away, CMS assumes it was due to
+reordering. Otherwise it is a genuine x86 instruction and gets propogated up to
+the program.
+
+Interrupts work similarly to exceptions, but CMS does not try retranslating the
+region in which the interrupt occurs.
+
+### Reordering constraints
+Consider a program that writes a 1 to address `x` and then reads from address
+`y`. If these instructions are reordered and `x = y`, then the resulting
+compiled program will not be correct. Similarly, if `x` and `y` are backed by
+a memory-mapped I/O device, the compiled program would have different I/O
+behavior than the original program.
+
+The CMS turns these potential issues into faults and handles them like
+exceptions. Reorderd instructions are tagged to let the processor know they were
+reordered. Special "alias hardware" does lightweight alias tracking at run time
+and faults if there may be aliasing between two reordered operations.
+Reordered operations that access IO address space also fault. The offending code
+region is then recompiled without the reorderings.
```

Comment:
Along the "magic keyword" lines above, it might be nice to acknowledge here that what they're doing is _speculation_, i.e., they're speculating that aliasing does not occur and that pointers do not point to memory-mapped IO.